src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -92,10 +93,22 @@ internal sealed partial class LocalRewriter : BoundTreeRewriterWithStackGuard
                 var localRewriter = new LocalRewriter(compilation, method, methodOrdinal, statement, containingType, factory, previousSubmissionFields, allowOmissionOfConditionalCalls, diagnostics,
                                                       dynamicInstrumenter != null ? new DebugInfoInjector(dynamicInstrumenter) : DebugInfoInjector.Singleton);
 
+                statement.CheckLocalsDefined();
                 var loweredStatement = (BoundStatement)localRewriter.Visit(statement);
+                loweredStatement.CheckLocalsDefined();
                 sawLambdas = localRewriter._sawLambdas;
                 sawLocalFunctions = localRewriter._sawLocalFunctions;
                 sawAwaitInExceptionHandler = localRewriter._sawAwaitInExceptionHandler;
+
+                if (localRewriter._needsSpilling && !loweredStatement.HasErrors)
+                {
+                    // Move spill sequences to a top-level statement. This handles "lifting" await and the switch expression.
+                    // PROTOTYPE(patterns2): should this be something that the caller does? It isn't really a "local" rewrite.
```

Comment:
I think it is better be handled by the caller, similarly to handling `await` in exception handling regions. 
- to have clear sequence of tree transformation. 
- to have better attribution when investigating perf impact of lowering passes.

Reply:
I placed this here because there is more than one caller. However We could duplicate this code for each caller. I am not sure which is better, but this simplifies the job of the caller.

---
In reply to: [178200828](https://github.com/dotnet/roslyn/pull/25697#discussion_r178200828) [](ancestors = 178200828)